### PR TITLE
feat(bigquery): Add Range Partitioning

### PR DESCRIPTION
**Pull request by quartzmo**
* Add range partitioning methods to `Table` and `Table::Updater`
* Add range partitioning methods to `LoadJob`
* Add range partitioning methods to `QueryJob`

closes: #4781

### CODE REVIEW

File: google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb [hunk 1]
```diff
@@ -555,6 +555,40 @@ def delete force: nil
         #     end
         #   end
         #
+        # @example With time partitioning and clustering.
+        #   require "google/cloud/bigquery"
+        #
+        #   bigquery = Google::Cloud::Bigquery.new
+        #   dataset = bigquery.dataset "my_dataset"
+        #
+        #   table = dataset.create_table "my_table" do |table|
```
Comment (dazuma) [inline google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb]: I find it confusing to have a block variable shadow an outer variable of the same name. Can you rename one of the variables? (Same for the other example below.)

File: google-cloud-bigquery/lib/google/cloud/bigquery/load_job.rb [hunk 2]
```diff
@@ -348,7 +348,71 @@ def output_bytes
         end
 
         ###
-        # Checks if the destination table will be time-partitioned. See
+        # Checks if the destination table will be range partitioned. See [Creating and using integer range partitioned
+        # tables](https://cloud.google.com/bigquery/docs/creating-integer-range-partitions).
+        #
+        # @return [Boolean, nil] `true` when the table is range partitioned, or `false` otherwise.
```
Comment (dazuma) [inline google-cloud-bigquery/lib/google/cloud/bigquery/load_job.rb]: When will it be `nil`?

File: google-cloud-bigquery/lib/google/cloud/bigquery/load_job.rb [hunk 3]
```diff
@@ -1253,14 +1317,180 @@ def labels= val
...
-          # Sets the partitioning for the destination table. See [Partitioned
+          # Sets the field on which to range partition the table. See [Creating and using integer range partitioned
+          # tables](https://cloud.google.com/bigquery/docs/creating-integer-range-partitions).
+          #
+          # See {#range_partitioning_start=}, {#range_partitioning_interval=} and {#range_partitioning_end=}.
+          #
+          # You can only set range partitioning when creating a table. BigQuery does not allow you to change
+          # partitioning on an existing table.
+          #
+          # @param [Integer] field The range partition field. the destination table is partitioned by this
```
Comment (dazuma) [inline google-cloud-bigquery/lib/google/cloud/bigquery/load_job.rb]: I think this type is supposed to be String, for the field name.

File: google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb [hunk 4]
```diff
@@ -410,6 +410,70 @@ def encryption
           EncryptionConfiguration.from_gapi @gapi.configuration.query.destination_encryption_configuration
         end
 
+        ###
+        # Checks if the destination table will be range partitioned. See [Creating and using integer range partitioned
+        # tables](https://cloud.google.com/bigquery/docs/creating-integer-range-partitions).
+        #
+        # @return [Boolean, nil] `true` when the table is range partitioned, or `false` otherwise.
```
Comment (dazuma) [inline google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb]: Nil?

File: google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb [hunk 5]
```diff
@@ -410,6 +410,70 @@ def encryption
...
+        # tables](https://cloud.google.com/bigquery/docs/creating-integer-range-partitions).
+        #
+        # @return [Integer, nil] The width of each interval, for data in range partitions, or `nil` if not range
+        #   partitioned.
+        #
+        # @!group Attributes
+        #
+        def range_partitioning_interval
+          return nil unless range_partitioning?
+          @gapi.configuration.query.range_partitioning.range.interval
```
Comment (dazuma) [inline google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb]: Any reason why this implementation is different from the other methods? why not:
```ruby
@gapi.configuration.query.range_partitioning.range.interval if range_partitioning?
```

File: google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb [hunk 6]
```diff
@@ -1028,6 +1092,164 @@ def encryption= val
...
+          ##
+          # Sets the field on which to range partition the table. See [Creating and using integer range partitioned
+          # tables](https://cloud.google.com/bigquery/docs/creating-integer-range-partitions).
+          #
+          # See {#range_partitioning_start=}, {#range_partitioning_interval=} and {#range_partitioning_end=}.
+          #
+          # You can only set range partitioning when creating a table. BigQuery does not allow you to change
+          # partitioning on an existing table.
+          #
+          # @param [Integer] field The range partition field. the destination table is partitioned by this
```
Comment (dazuma) [inline google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb]: String

File: google-cloud-bigquery/lib/google/cloud/bigquery/table.rb [hunk 7]
```diff
@@ -2521,6 +2606,168 @@ def initialize gapi
...
+          # Sets the field on which to range partition the table. See [Creating and using integer range partitioned
+          # tables](https://cloud.google.com/bigquery/docs/creating-integer-range-partitions).
+          #
+          # See {Table::Updater#range_partitioning_start=}, {Table::Updater#range_partitioning_interval=} and
+          # {Table::Updater#range_partitioning_end=}.
+          #
+          # You can only set range partitioning when creating a table as in the example below. BigQuery does not allow
+          # you to change partitioning on an existing table.
+          #
+          # @param [Integer] field The range partition field. The table is partitioned by this
```
Comment (dazuma) [inline google-cloud-bigquery/lib/google/cloud/bigquery/table.rb]: String

File: google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb (hunk 1 above)
Comment (quartzmo) [reply google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb]: This was done in 11 places. I'll change them all to `table = dataset.create_table "my_table" do |t|`.

File: google-cloud-bigquery/lib/google/cloud/bigquery/table.rb (hunk 7 above)
Comment (quartzmo) [reply google-cloud-bigquery/lib/google/cloud/bigquery/table.rb]: Fixed, thanks!

File: google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb (hunk 6 above)
Comment (quartzmo) [reply google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb]: Fixed, thanks!

File: google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb (hunk 5 above)
Comment (quartzmo) [reply google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb]: Fixed, thanks!

File: google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb (hunk 4 above)
Comment (quartzmo) [reply google-cloud-bigquery/lib/google/cloud/bigquery/query_job.rb]: Fixed, thanks!

File: google-cloud-bigquery/lib/google/cloud/bigquery/load_job.rb (hunk 3 above)
Comment (quartzmo) [reply google-cloud-bigquery/lib/google/cloud/bigquery/load_job.rb]: Fixed, thanks!

File: google-cloud-bigquery/lib/google/cloud/bigquery/load_job.rb (hunk 2 above)
Comment (quartzmo) [reply google-cloud-bigquery/lib/google/cloud/bigquery/load_job.rb]: Fixed, thanks!

File: google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb (hunk 1 above)
Comment (quartzmo) [reply google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb]: Updated.

**Pull request merged**